### ISSUE TITLE
Change the token used to access Che tenant namespace in OSIO

### ISSUE THREAD

**Issue by l0rd** [redhat-developer/rh-che#532]
We currently use users' tokens to create Che workspaces in Che tenant namespace. We need to change that flow to:

At startup:
- [x] Use an openshift secret to login to auth service - PR is sent https://github.com/redhat-developer/rh-che/pull/538
- [x] Get che token from auth service - PR is sent https://github.com/redhat-developer/rh-che/pull/538

When creating/stopping/idling a workspace
- [x] Get current user ID from user jwt token - PR is sent https://github.com/redhat-developer/rh-che/pull/533
- [x] Use che token + user ID (as a query param) for proxy requests (auth will return a Che ServiceAccount token)

And this should be deployed using a canary rollout using unleash. So we will also need to:

- [x] Add unleash support in rh-che - PR is sent https://github.com/redhat-developer/rh-che/pull/537

**Comment by sbose78** [comment]
Support for Che to 'login' and get an OSiO service account token tracked in https://github.com/fabric8-services/fabric8-auth/issues/316

**Comment by ibuziuk** [comment]
Some updates related to:

```
At startup:
- [ ] Use an openshift secret to login to auth service
- [ ] Get che token from auth service
```
Housekeeping issue[1] is closed and now `service.account.id` & `service.account.secret` are added to che secret in prod + preview. What is required from che side is:
- add `service.account.id` / `service.account.secret` as env vars to template[2]
- obtain Che SA token via `POST /api/token` [3] where values provided in env vars are used in request parameters

[1] https://gitlab.cee.redhat.com/dtsd/housekeeping/issues/1222#note_264499 
[2] https://github.com/redhat-developer/rh-che/blob/master/openshift/rh-che.app.yaml#L246
[3] https://fabric8-services.github.io/fabric8-auth/reference.html#_service_account_authentication

**Comment by sbose78** [comment]
Auth service is prod-preview and prod now supports service account login from Che.

**Comment by alexeykazakov** [comment]
Not in prod yet.
We need to redeploy prod to pick up the new configuration.

**Comment by ibuziuk** [comment]
PR for adding  che SA to 'fabric8-tenant-che-mt' - https://github.com/fabric8-services/fabric8-tenant-che/pull/97
PR for adding fabric8-auth endpoint to template -https://github.com/redhat-developer/rh-che/pull/535

@davidfestal am I correct that the migration job does not remove service accounts [1]  ?

https://github.com/fabric8-services/fabric8-tenant-che-migration/blob/master/source/io/fabric8/tenant/che/migration/namespace/cleanup.ceylon#L119-L126

**Comment by davidfestal** [comment]
> @davidfestal am I correct that the migration job does not remove service accounts [1] ?

no, it doesn't remove service accounts

**Comment by ibuziuk** [comment]
@aslakknutsen do you happen to know the status of the PR for oso-proxy was it already merged (could you please link it to this issue) ? We are planning to proceed with canary rollout of che namespace lookdown this sprint

**Comment by garagatyi** [comment]
@ibuziuk we still need to send user ID to OSO proxy. But we would be able to do that only after proper changes you are talking about

**Comment by ibuziuk** [comment]
@garagatyi sure thing, that is the plan ;-)

**Comment by aslakknutsen** [comment]
There are 3 changes ready/coming to proxy this week. The current plan is to roll out next Monday when I'm back from PTO.

**Comment by ibuziuk** [comment]
@aslakknutsen am I correct that oso-proxy is now supporting Che SA token [1]. Could you please tell what query parameter is expected from che ? 

[1] https://github.com/fabric8-services/fabric8-oso-proxy/commit/c3ecdb11a29b5d9dbda929e037bbe1c8946173bc

**Comment by ibuziuk** [comment]
che sa token support in rh-che / oso-proxy is already available on prod-preview and seems to be working just fine. All relevant changes have been done on rh-che & oso-proxy sides. I have created a separate issue [1] for switching to the `Impersonate-User` header. Assuming that oso-proxy currently support both approaches of identity_id processing (header and query parameter) this new issue should be treated as an internal implementation improvement which does not block *-che namespace lockdown and does not affect the end user.

[1] https://github.com/redhat-developer/rh-che/issues/703
